src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray_1.cs

```diff
@@ -575,17 +575,8 @@ public ImmutableArray<T> RemoveRange(IEnumerable<T> items, IEqualityComparer<T>?
             self.ThrowNullRefIfNotInitialized();
             Requires.NotNull(items, nameof(items));
 
-            var indicesToRemove = new SortedSet<int>();
-            foreach (T item in items)
-            {
-                int index = -1;
-                do
-                {
-                    index = self.IndexOf(item, index + 1, equalityComparer);
-                } while (index >= 0 && !indicesToRemove.Add(index) && index < self.Length - 1);
-            }
-
-            return self.RemoveAtRange(indicesToRemove);
+            ICollection<int>? indicesToRemove = items.TryGetCount(out _) ? self.FindOrderedIndicesToRemoveByMultiSet(items, equalityComparer) : self.FindOrderedIndicesToRemove(items, equalityComparer);
```

Comment:
I'm not sure I understand why we need to fall back to the old algorithm if we can't determine the size?

Reply:
See previous [comment(61196)](https://github.com/dotnet/runtime/pull/61196#discussion_r793513803)
Here we use `TryGetCount` to test type.

Reply:
Maybe add a comment explaining this? What about collections with very large counts?

Reply:
> Maybe add a comment explaining this?

Added.

> What about collections with very large counts?

Not sure, I just feel the fall back code is for true enumerable data which sounds like long `streaming` concept. But if it is already a large reality collection data, that will mean that it was successful to allocate large memory (so same order of magnitude). Please give suggestion.

Reply:
One possible approach might be to fall back to the slow implementation once the multiset exceeds a certain capacity dynamically:
```C#
        private ICollection<int>? FindOrderedIndicesToRemoveByMultiSet(IEnumerable<T> items, IEqualityComparer<T>? equalityComparer)
        {
            var multiSet = new MultiSet(equalityComparer);
            int i = 0;
            foreach (T item in items)
            {
                multiSet.Add(item);
                if (++i == 100) // threshold way above what I would expect is the typical input enumerable.
                {
                      return FindOrderedIndicesToRemove(items, equalityComparer);
                }
            }

            return GetOrderedIndicesToRemoveFor(multiSet);
        }
```
I would strongly encourage you to make any decision based on benchmark numbers, however.

Reply:
// Not check size
           public ICollection<int>? Fast_NoLimit_FindOrderedIndicesToRemoveByMultiSet(IEnumerable<T> items, IEqualityComparer<T>? equalityComparer)
            {
                var multiSet = new MultiSet(equalityComparer);
                foreach (T item in items)
                {
                    multiSet.Add(item);
                }

                return GetOrderedIndicesToRemoveFor(multiSet);
            }

          // Firstly try to get collection count property directly then fall back to iterate to trigger size limit dynamically
           public ICollection<int>? Fast_TryGetCountThenIterateCheckCount(IEnumerable<T> items, IEqualityComparer<T>? equalityComparer)
            {
                if (items.TryGetCount(out int count))
                {
                    if (count <= ThrottleValue)
                    {
                        return Fast_NoLimit_FindOrderedIndicesToRemoveByMultiSet(items, equalityComparer);
                    }

                    return SlowFindOrderedIndicesToRemove(items, equalityComparer);
                }

                return Fast_IterateCheckCount(items, equalityComparer);
            }

          // Firstly try to get collection count property directly then fall back to iterate to trigger size limit compared with actual dictionary.count (dedup same value purpose) dynamically
            public ICollection<int>? Fast_TryGetCountThenIterateCheckCountByDicLength(IEnumerable<T> items, IEqualityComparer<T>? equalityComparer)
            {
                if (items.TryGetCount(out int count))
                {
                    if (count <= ThrottleValue)
                    {
                        return Fast_NoLimit_FindOrderedIndicesToRemoveByMultiSet(items, equalityComparer);
                    }

                    return SlowFindOrderedIndicesToRemove(items, equalityComparer);
                }

                return Fast_IterateCheckCountByDicLength(items, equalityComparer);
            }

          // Iterate to trigger size limit compared with actual dictionary.count (dedup same value purpose) dynamically
            public ICollection<int>? Fast_IterateCheckCountByDicLength(IEnumerable<T> items, IEqualityComparer<T>? equalityComparer)
            {
                var multiSet = new MultiSet(equalityComparer);
                foreach (T item in items)
                {
                    if (multiSet.ItemCount > ThrottleValue)
                    {
                        return SlowFindOrderedIndicesToRemove(items, equalityComparer);
                    }
                    multiSet.Add(item);
                }

                return GetOrderedIndicesToRemoveFor(multiSet);
            }

          // Iterate to trigger size limit dynamically
            public ICollection<int>? Fast_IterateCheckCount(IEnumerable<T> items, IEqualityComparer<T>? equalityComparer)
            {
                var multiSet = new MultiSet(equalityComparer);
                int i = 0;
                foreach (T item in items)
                {
                    if (++i > ThrottleValue)
                    {
                        return SlowFindOrderedIndicesToRemove(items, equalityComparer);
                    }
                    multiSet.Add(item);
                }

                return GetOrderedIndicesToRemoveFor(multiSet);
            }

           private readonly struct MultiSet
           {
                public int ItemCount => _dictionary.Count;
           }

Benchmark condition:
ImmutableArray length = 100, Items length = 100

1. `Items `is true enumerable
![image](https://user-images.githubusercontent.com/72254037/156095858-64f2c2f7-371a-47db-a5d5-9563754d2e6b.png)
2. `Items` is array instance
![image](https://user-images.githubusercontent.com/72254037/156095934-dc08847a-8817-469a-8c2a-de6df991f9ee.png)

Result:
Pure fast method is ~1x faster than slow method
If not hit `throttleValue`, pure fast is almost same time as all other fast methods with throttling check
If hit `throttleValue` so that fall back to slow path, then all fast methods with throttling check are much slower than original slow method, except "`Fast_TryGetCountThenIterateCheckCount()`" and "`Fast_TryGetCountThenIterateCheckCountByDicLength()`" when `items` is Array instance because they can early find situation to fall back to slow path. However they are still slower than original slow method.

So if we need to have size limit, maybe `Fast_TryGetCountThenIterateCheckCountByDicLength()` is better, but not sure what throttling value is better, LOH triggering size is 85k but need to find out it after Dictionary already scales array size (`EnsureCapacity`). Need guidance for that, thanks.
Otherwise, pure fast method can give shortest time in all cases.

Reply:
Thanks, I think `Fast_TryGetCountThenIterateCheckCountByDicLength` is ok, although I would be interested in seeing the numbers for various sizes of input enumerables, particularly for low values like 1, 2 or 10 elements. It might be the case that for low enough values creating a MultiSet actually makes things slower. You can execute benchmarks for multiple sizes using BDN's `ParamsAttribute`.

Reply:
Under condition of 100 length ImmutableArray, fast method is slower until items length up to ~30.
![image](https://user-images.githubusercontent.com/72254037/156381903-e9e19378-0f1a-4255-9061-8a019bb6394d.png)
![image](https://user-images.githubusercontent.com/72254037/156382170-0e0f0b2e-ca60-43e0-9fee-33bbace2705e.png)

Reply:
Hmmm, these numbers make me think that this optimization is likely not worth our while. I don't think it's worth optimizing for inputs > 30, it's unlikely that this method would be used with so large inputs. I'd probably just close this PR. Thanks for driving the performance investigation!